File: course_3/module_2/examples/union_find.go
package main

import (
	"errors"
	"fmt"
)

// UnionFind represents the Union-Find structure using Lazy Unions, Union by Rank, and Path Compression.
type UnionFind struct {
	parent []int // Parent pointers for each element.
	rank   []int // Rank of each set (helps optimize unions).
}

// NewUnionFind initializes a new Union-Find structure with the given size.
// Each element starts in its own set, where it is its own parent and rank is 0.
func NewUnionFind(size int) *UnionFind {
	if size <= 0 {
		panic("Size must be greater than 0.")
	}
	parent := make([]int, size)
	rank := make([]int, size)
	for i := 0; i < size; i++ {
		parent[i] = i // Each element is its own parent.
		rank[i] = 0   // Initial rank for all elements is 0.
	}
	return &UnionFind{parent: parent, rank: rank}
}

// validate ensures that the index x is within the valid range.
func (uf *UnionFind) validate(x int) error {
	if x < 0 || x >= len(uf.parent) {
		return errors.New(fmt.Sprintf("Index %d is out of bounds.", x))
	}
	return nil
}

// Find returns the root of the element x, applying path compression.
// Path compression ensures that all elements along the path point directly to the root,
// flattening the tree and optimizing future operations.
func (uf *UnionFind) Find(x int) (int, error) {
	if err := uf.validate(x); err != nil {
		return -1, err
	}
	if uf.parent[x] != x {
		root, err := uf.Find(uf.parent[x])
		if err != nil {
			return -1, err
		}
		uf.parent[x] = root // Path compression.
	}
	return uf.parent[x], nil
}

// Union merges the sets containing x and y using union by rank.
// The root of the smaller rank tree is made a child of the root of the larger rank tree.
// If ranks are equal, one root is chosen arbitrarily, and its rank is incremented.
func (uf *UnionFind) Union(x, y int) error {
	rootX, err := uf.Find(x)
	if err != nil {
		return err
	}
	rootY, err := uf.Find(y)
	if err != nil {
		return err
	}

	if rootX != rootY {
		// Merge smaller rank tree into the larger rank tree.
		if uf.rank[rootX] < uf.rank[rootY] {
			uf.parent[rootX] = rootY
		} else if uf.rank[rootX] > uf.rank[rootY] {
			uf.parent[rootY] = rootX
		} else {
			// If ranks are equal, choose one as root and increment its rank.
			uf.parent[rootY] = rootX
			uf.rank[rootX]++
		}
	}
	return nil
}

// Connected checks if two elements x and y are in the same set.
// Returns true if they share the same root, false otherwise.
func (uf *UnionFind) Connected(x, y int) (bool, error) {
	rootX, err := uf.Find(x)
	if err != nil {
		return false, err
	}
	rootY, err := uf.Find(y)
	if err != nil {
		return false, err
	}
	return rootX == rootY, nil
}

// Components returns all connected components as a map.
// The keys of the map are the roots, and the values are slices of elements in each component.
func (uf *UnionFind) Components() map[int][]int {
	components := make(map[int][]int)
	visited := make(map[int]int) // Cache the root of each element.
	for i := 0; i < len(uf.parent); i++ {
		if root, found := visited[i]; found {
			components[root] = append(components[root], i)
		} else {
			root, _ := uf.Find(i) // We already validated indices, no need for extra checks.
			visited[i] = root
			components[root] = append(components[root], i)
		}
	}
	return components
}

// Size returns the number of elements in the Union-Find structure.
func (uf *UnionFind) Size() int {
	return len(uf.parent)
}

// Main function to demonstrate usage of the Union-Find structure.
func main() {
	uf := NewUnionFind(10)

	// Perform union operations.
	_ = uf.Union(1, 2)
	_ = uf.Union(3, 4)
	_ = uf.Union(2, 4)

	// Check connectivity.
	connected, _ := uf.Connected(1, 3)
	fmt.Println("Connected(1, 3):", connected)
	connected, _ = uf.Connected(1, 5)
	fmt.Println("Connected(1, 5):", connected)

	// Print all connected components.
	fmt.Println("Components:")
	components := uf.Components()
	for root, group := range components {
		fmt.Printf("Root %d: %v\n", root, group)
	}
}

File: course_3/module_2/examples/union_find.py
class UnionFind:
    def __init__(self, size: int):
        """
        Initialize the Union-Find structure with Lazy Unions, Union by Rank, and Path Compression.
        :param size: Number of elements.
        """
        if size <= 0:
            raise ValueError("Size must be greater than 0.")
        self.parent = list(range(size))  # Parent pointers.
        self.rank = [0] * size           # Rank of each set.

    def find(self, x: int) -> int:
        """
        Find the root of the element x with path compression.
        :param x: Element to find.
        :return: Root of the element.
        """
        if x < 0 or x >= len(self.parent):
            raise ValueError(f"Index {x} is out of bounds.")
        if self.parent[x] != x:
            # Path compression: update parent to the root.
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int):
        """
        Union by rank of the sets containing x and y.
        :param x: First element.
        :param y: Second element.
        """
        root_x = self.find(x)
        root_y = self.find(y)

        if root_x != root_y:
            # Union by rank: attach smaller tree under larger tree.
            if self.rank[root_x] < self.rank[root_y]:
                self.parent[root_x] = root_y
            elif self.rank[root_x] > self.rank[root_y]:
                self.parent[root_y] = root_x
            else:
                # If ranks are equal, choose one as root and increment its rank.
                self.parent[root_y] = root_x
                self.rank[root_x] += 1

    def connected(self, x: int, y: int) -> bool:
        """
        Check if two elements are in the same set.
        :param x: First element.
        :param y: Second element.
        :return: True if they are in the same set, False otherwise.
        """
        return self.find(x) == self.find(y)

    def components(self) -> dict:
        """
        Get all connected components as a dictionary.
        :return: Dictionary of root to list of elements.
        """
        groups = {}
        for i in range(len(self.parent)):
            root = self.find(i)
            if root not in groups:
                groups[root] = []
            groups[root].append(i)
        return groups


# Example usage:
if __name__ == "__main__":
    uf = UnionFind(10)

    # Perform union operations.
    uf.union(1, 2)
    uf.union(3, 4)
    uf.union(2, 4)

    # Check connectivity.
    print(uf.connected(1, 3))
    print(uf.connected(1, 5))

    # Print all components in the desired format.
    print("Components:")
    components = uf.components()
    for root, group in components.items():
        print(f"Root {root}: {group}")
